helpers/python

```diff
@@ -3,12 +3,135 @@
 import pkgutil
 import sys
 
-# walk_packages() is much slower than iter_modules(), use it only when
-# completing something with a dot in it.
-if len(sys.argv) > 1 and "." in sys.argv[1]:
-    walker = pkgutil.walk_packages
-else:
-    walker = pkgutil.iter_modules
-
-for mod in walker():
-    print(mod[1])  # noqa: E211
+
+class PackageNotFoundError(Exception):
+    def __init__(self, name):
+        Exception.__init__(self, name)
+        self.name = name
+
+
+try:
+    from importlib import import_module
+except ImportError:
+
+    def import_module(name):
+        __import__(name)
+        return sys.modules[name]
+
+
+try:
+    from importlib.util import find_spec
+except ImportError:
+    find_spec = None
+
+
+def print_error(message):
+    sys.stderr.write("error: %s\n" % message)
+
+
+def import_package(pkgname):
+    """Import a module only if it is known to be a package.
+
+    The purpose is to reduce the risk of unwanted side effects caused by
+    importing misbehaving modules (e.g. ones that are designed to be only run
+    as programs but lack `__name__ == "__main__"` guard). Thus, already loaded
+    modules are returned as is even if they are not packages.
+    """
+    pkg = sys.modules.get(pkgname)
+    if pkg is not None:
+        return pkg
+    parentname = pkgname.rpartition(".")[0]
+    if parentname:
+        parent = import_package(parentname)
+    if find_spec is not None:
+        spec = find_spec(pkgname)
+        if spec is not None and spec.submodule_search_locations is not None:
+            return import_module(pkgname)
+    # find_spec() unavailable, fall back to searching iter_modules(...).
+    if parentname:
+        path = parent.__path__ or []
```

Comment:
`__path__` is documented as
> A list of strings specifying the search path within a package. This attribute is not set on modules.

Can we take advantage of that when deciding if something is (not) a package?